diff --git a/Makefile.def b/Makefile.def
index 5116341..1f27ac5 100644
--- a/Makefile.def
+++ b/Makefile.def
@@ -47,7 +47,7 @@ host_modules= { module= gas; bootstrap=true; };
 host_modules= { module= gcc; bootstrap=true; 
 		extra_make_flags="$(EXTRA_GCC_FLAGS)"; };
 host_modules= { module= gmp; lib_path=.libs; bootstrap=true;
-		extra_configure_flags='--disable-shared';
+		extra_configure_flags='--disable-shared @extra_gmp_configure_flags@';
 		no_install= true;
 		// none-*-* disables asm optimizations, bootstrap-testing
 		// the compiler more thoroughly.
@@ -62,12 +62,12 @@ host_modules= { module= mpc; lib_path=.libs; bootstrap=true;
 		extra_configure_flags='--disable-shared @extra_mpc_gmp_configure_flags@ @extra_mpc_mpfr_configure_flags@';
 		no_install= true; };
 host_modules= { module= ppl; lib_path=src/.libs; bootstrap=true;
-		extra_configure_flags='--disable-shared --with-libgmp-prefix=$$r/$(HOST_SUBDIR)/gmp/ --with-libgmpxx-prefix=$$r/$(HOST_SUBDIR)/gmp/';
+		extra_configure_flags='--disable-shared --with-gmp-build=$$r/$(HOST_SUBDIR)/gmp/ CPPFLAGS="-I$$r/$(HOST_SUBDIR)/gmp -I$${topdir}/gmp/"';
 		no_install= true; };
 host_modules= { module= cloog; lib_path=.libs; bootstrap=true;
 		extra_configure_flags='--disable-shared --with-gmp-library=$$r/$(HOST_SUBDIR)/gmp/.libs --with-gmp-include=$$r/$(HOST_SUBDIR)/gmp --with-bits=gmp --with-ppl';
-		extra_exports='CPPFLAGS="${CPPFLAGS} ${HOST_PPLINC}"; export CPPFLAGS; LDFLAGS="$$LDFLAGS ${HOST_PPLLIBS}"; export LDFLAGS; ';
-		extra_make_flags='CPPFLAGS="${CPPFLAGS} ${HOST_PPLINC}" LDFLAGS="$$LDFLAGS"';
+		extra_exports='CPPFLAGS="${CPPFLAGS} -I$$r/$(HOST_SUBDIR)/gmp ${HOST_PPLINC}"; export CPPFLAGS; LDFLAGS="$$LDFLAGS -L$$r/$(HOST_SUBDIR)/gmp/.libs ${HOST_PPLLIBS}"; export LDFLAGS; ';
+		extra_make_flags='CPPFLAGS="${CPPFLAGS} -I$$r/$(HOST_SUBDIR)/gmp ${HOST_PPLINC}" LDFLAGS="$$LDFLAGS -L$$r/$(HOST_SUBDIR)/gmp/.libs"';
 		no_install= true; };
 host_modules= { module= libelf; lib_path=.libs; bootstrap=true;
 		extra_configure_flags='--disable-shared';
diff --git a/Makefile.in b/Makefile.in
index d1206bd..5aacac2 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -10368,7 +10368,7 @@ configure-gmp:
 	libsrcdir="$$s/gmp"; \
 	$(SHELL) $${libsrcdir}/configure \
 	  $(HOST_CONFIGARGS) --build=${build_alias} --host=none-${host_vendor}-${host_os} \
-	  --target=none-${host_vendor}-${host_os} $${srcdiroption} --disable-shared \
+	  --target=none-${host_vendor}-${host_os} $${srcdiroption} --disable-shared --enable-cxx \
 	  || exit 1
 @endif gmp
 
@@ -12993,7 +12993,7 @@ configure-ppl:
 	libsrcdir="$$s/ppl"; \
 	$(SHELL) $${libsrcdir}/configure \
 	  $(HOST_CONFIGARGS) --build=${build_alias} --host=${host_alias} \
-	  --target=${target_alias} $${srcdiroption} --disable-shared --with-libgmp-prefix=$$r/$(HOST_SUBDIR)/gmp/ --with-libgmpxx-prefix=$$r/$(HOST_SUBDIR)/gmp/ \
+	  --target=${target_alias} $${srcdiroption} --disable-shared --with-gmp-build=$$r/$(HOST_SUBDIR)/gmp/ CPPFLAGS="-I$$r/$(HOST_SUBDIR)/gmp -I$${topdir}/gmp/" \
 	  || exit 1
 @endif ppl
 
@@ -13856,7 +13856,7 @@ configure-cloog:
 	s=`cd $(srcdir); ${PWD_COMMAND}`; export s; \
 	test ! -f $(HOST_SUBDIR)/cloog/Makefile || exit 0; \
 	$(SHELL) $(srcdir)/mkinstalldirs $(HOST_SUBDIR)/cloog ; \
-	$(HOST_EXPORTS) CPPFLAGS="${CPPFLAGS} ${HOST_PPLINC}"; export CPPFLAGS; LDFLAGS="$$LDFLAGS ${HOST_PPLLIBS}"; export LDFLAGS;  \
+	$(HOST_EXPORTS) CPPFLAGS="${CPPFLAGS} -I$$r/$(HOST_SUBDIR)/gmp ${HOST_PPLINC}"; export CPPFLAGS; LDFLAGS="$$LDFLAGS -L$$r/$(HOST_SUBDIR)/gmp/.libs ${HOST_PPLLIBS}"; export LDFLAGS;  \
 	echo Configuring in $(HOST_SUBDIR)/cloog; \
 	cd "$(HOST_SUBDIR)/cloog" || exit 1; \
 	case $(srcdir) in \
@@ -13868,7 +13868,7 @@ configure-cloog:
 	libsrcdir="$$s/cloog"; \
 	$(SHELL) $${libsrcdir}/configure \
 	  $(HOST_CONFIGARGS) --build=${build_alias} --host=${host_alias} \
-	  --target=${target_alias} $${srcdiroption} --disable-shared --with-gmp-library=$$r/$(HOST_SUBDIR)/gmp/.libs --with-gmp-include=$$r/$(HOST_SUBDIR)/gmp --with-bits=gmp --with-ppl \
+	  --target=${target_alias} $${srcdiroption} --disable-shared LDFLAGS="-L$$r/$(HOST_SUBDIR)/gmp/.libs" CPPFLAGS="-I$$r/$(HOST_SUBDIR)/gmp" --with-bits=gmp --with-ppl \
 	  || exit 1
 @endif cloog
 
@@ -14093,7 +14093,7 @@ all-cloog: configure-cloog
 	s=`cd $(srcdir); ${PWD_COMMAND}`; export s; \
 	$(HOST_EXPORTS) CPPFLAGS="${CPPFLAGS} ${HOST_PPLINC}"; export CPPFLAGS; LDFLAGS="$$LDFLAGS ${HOST_PPLLIBS}"; export LDFLAGS;  \
 	(cd $(HOST_SUBDIR)/cloog && \
-	  $(MAKE) $(BASE_FLAGS_TO_PASS) $(EXTRA_HOST_FLAGS) CPPFLAGS="${CPPFLAGS} ${HOST_PPLINC}" LDFLAGS="$$LDFLAGS" \
+	  $(MAKE) $(BASE_FLAGS_TO_PASS) $(EXTRA_HOST_FLAGS) CPPFLAGS="${CPPFLAGS} -I$$r/$(HOST_SUBDIR)/gmp ${HOST_PPLINC}" LDFLAGS="$$LDFLAGS -L$$r/$(HOST_SUBDIR)/gmp/.libs" \
 		$(TARGET-cloog))
 @endif cloog
 
diff --git a/configure b/configure
index eb69f28..19015dc 100755
--- a/configure
+++ b/configure
@@ -5164,9 +5164,9 @@ if test "x$with_mpfr_lib" != x; then
   gmplibs="-L$with_mpfr_lib $gmplibs"
 fi
 if test "x$with_mpfr$with_mpfr_include$with_mpfr_lib" = x && test -d ${srcdir}/mpfr; then
-  gmplibs='-L$$r/$(HOST_SUBDIR)/mpfr/'"$lt_cv_objdir $gmplibs"
-  gmpinc='-I$$r/$(HOST_SUBDIR)/mpfr -I$$s/mpfr '"$gmpinc"
-  extra_mpc_mpfr_configure_flags='--with-mpfr-include=$$s/mpfr --with-mpfr-lib=$$r/$(HOST_SUBDIR)/mpfr/'"$lt_cv_objdir"
+  gmplibs='-L$$r/$(HOST_SUBDIR)/mpfr/src/'"$lt_cv_objdir $gmplibs"
+  gmpinc='-I$$r/$(HOST_SUBDIR)/mpfr/src -I$$s/mpfr/src '"$gmpinc"
+  extra_mpc_mpfr_configure_flags='--with-mpfr-include=$$s/mpfr/src --with-mpfr-lib=$$r/$(HOST_SUBDIR)/mpfr/src/'"$lt_cv_objdir"
   # Do not test the mpfr version.  Assume that it is sufficient, since
   # it is in the source tree, and the library has not been built yet
   # but it would be included on the link line in the version check below
@@ -5603,6 +5603,7 @@ if test x"$with_ppl$with_ppl_include$with_ppl_lib" = x && test -d ${srcdir}/ppl;
 fi
 
 if test "x$with_ppl" != xno; then
+  extra_gmp_configure_flags='--enable-cxx'
   if test "x$pwllib" = x; then
     saved_LIBS="$LIBS"
     LIBS="$LIBS $ppllibs -lstdc++ -lm"
diff --git a/configure.ac b/configure.ac
index 337e11d..48787d8 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1262,9 +1262,9 @@ if test "x$with_mpfr_lib" != x; then
   gmplibs="-L$with_mpfr_lib $gmplibs"
 fi
 if test "x$with_mpfr$with_mpfr_include$with_mpfr_lib" = x && test -d ${srcdir}/mpfr; then
-  gmplibs='-L$$r/$(HOST_SUBDIR)/mpfr/'"$lt_cv_objdir $gmplibs"
-  gmpinc='-I$$r/$(HOST_SUBDIR)/mpfr -I$$s/mpfr '"$gmpinc"
-  extra_mpc_mpfr_configure_flags='--with-mpfr-include=$$s/mpfr --with-mpfr-lib=$$r/$(HOST_SUBDIR)/mpfr/'"$lt_cv_objdir"
+  gmplibs='-L$$r/$(HOST_SUBDIR)/mpfr/src/'"$lt_cv_objdir $gmplibs"
+  gmpinc='-I$$r/$(HOST_SUBDIR)/mpfr/src -I$$s/mpfr/src '"$gmpinc"
+  extra_mpc_mpfr_configure_flags='--with-mpfr-include=$$s/mpfr/src --with-mpfr-lib=$$r/$(HOST_SUBDIR)/mpfr/src/'"$lt_cv_objdir"
   # Do not test the mpfr version.  Assume that it is sufficient, since
   # it is in the source tree, and the library has not been built yet
   # but it would be included on the link line in the version check below
@@ -1546,6 +1546,8 @@ if test "x$with_ppl" != xno; then
     LIBS="$saved_LIBS"
   fi
 
+  extra_gmp_configure_flags='--enable-cxx'
+
   ppllibs="$ppllibs -lppl_c -lppl $pwllib -lgmpxx"
 
   if test "$enable_ppl_version_check" != no; then
